### PR TITLE
Fixes bug when trimming DNS SRV records.

### LINKED ISSUE #1931: DNS Query for SRV Answer Truncated to Nothing
I was working with a small cluster this evening and registered 9 redis machines into my consul services catalog. When I went to dig for the `SRV` records, however, I noticed that I did not receive any answer section:

```
$ dig @127.0.0.1 -p 8600 redis-cache-redis.service.consul SRV

; <<>> DiG 9.9.5-3ubuntu0.6-Ubuntu <<>> @127.0.0.1 -p 8600 redis-cache-redis.service.consul SRV
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 52069
;; flags: qr aa rd; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 9
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;redis-cache-redis.service.consul. IN   SRV

;; AUTHORITY SECTION:
consul.                 0       IN      SOA     ns.consul. postmaster.consul. 1460264240 3600 600 86400 0

;; ADDITIONAL SECTION:
ip-10-0-1-185.node.dc1.consul. 0 IN     A       10.0.1.185
ip-10-0-1-16.node.dc1.consul. 0 IN      A       10.0.1.16
ip-10-0-1-59.node.dc1.consul. 0 IN      A       10.0.1.59
ip-10-0-1-59.node.dc1.consul. 0 IN      A       10.0.1.59
ip-10-0-1-185.node.dc1.consul. 0 IN     A       10.0.1.185
ip-10-0-1-59.node.dc1.consul. 0 IN      A       10.0.1.59
ip-10-0-1-16.node.dc1.consul. 0 IN      A       10.0.1.16
ip-10-0-1-185.node.dc1.consul. 0 IN     A       10.0.1.185
ip-10-0-1-16.node.dc1.consul. 0 IN      A       10.0.1.16

;; Query time: 5 msec
;; SERVER: 127.0.0.1#8600(127.0.0.1)
;; WHEN: Sun Apr 10 04:57:20 UTC 2016
;; MSG SIZE  rcvd: 517
```

All of those machines (nodes) in the `ADDITIONAL` section are correct. I did a bit of searching and found the `enable_truncate` flag, which I set, and got the same lack of answer when querying using UDP (however, now that the truncate flag was set, `dig` became smart enough to query using TCP and received the entire `ANSWER` section).

This leads me to the following question: is removing the _entire_ `ANSWER` section appropriate? I found the logic in the code that sees that the length of the message is too long and removes `ANSWER` entries until there are no more `ANSWER`s or it's under the size limit ([dns.go](https://github.com/hashicorp/consul/blob/master/command/agent/dns.go#L498)), but I don't know enough about the DNS and it's implementation to know the correct answer to that. It seems to me that completely removing the `ANSWER` section in favor of the `ADDITIONAL` section is bizarre. I would (maybe naïvely) expect both sections to be trimmed down the same hosts/entries so there a maximum, equal number of them.

Any thoughts?

For what it's worth, here's the response after `enable_truncate` is in place. Working perfectly now:

```
$ dig @127.0.0.1 -p 8600 redis-cache-redis.service.consul SRV
;; Truncated, retrying in TCP mode.

; <<>> DiG 9.9.5-3ubuntu0.6-Ubuntu <<>> @127.0.0.1 -p 8600 redis-cache-redis.service.consul SRV
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 11627
;; flags: qr aa rd; QUERY: 1, ANSWER: 9, AUTHORITY: 0, ADDITIONAL: 9
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;redis-cache-redis.service.consul. IN   SRV

;; ANSWER SECTION:
redis-cache-redis.service.consul. 0 IN  SRV     1 1 36518 ip-10-0-1-185.node.dc1.consul.
redis-cache-redis.service.consul. 0 IN  SRV     1 1 31562 ip-10-0-1-59.node.dc1.consul.
redis-cache-redis.service.consul. 0 IN  SRV     1 1 29888 ip-10-0-1-185.node.dc1.consul.
redis-cache-redis.service.consul. 0 IN  SRV     1 1 32688 ip-10-0-1-16.node.dc1.consul.
redis-cache-redis.service.consul. 0 IN  SRV     1 1 48075 ip-10-0-1-59.node.dc1.consul.
redis-cache-redis.service.consul. 0 IN  SRV     1 1 56648 ip-10-0-1-59.node.dc1.consul.
redis-cache-redis.service.consul. 0 IN  SRV     1 1 28523 ip-10-0-1-185.node.dc1.consul.
redis-cache-redis.service.consul. 0 IN  SRV     1 1 33573 ip-10-0-1-16.node.dc1.consul.
redis-cache-redis.service.consul. 0 IN  SRV     1 1 27304 ip-10-0-1-16.node.dc1.consul.

;; ADDITIONAL SECTION:
ip-10-0-1-185.node.dc1.consul. 0 IN     A       10.0.1.185
ip-10-0-1-59.node.dc1.consul. 0 IN      A       10.0.1.59
ip-10-0-1-185.node.dc1.consul. 0 IN     A       10.0.1.185
ip-10-0-1-16.node.dc1.consul. 0 IN      A       10.0.1.16
ip-10-0-1-59.node.dc1.consul. 0 IN      A       10.0.1.59
ip-10-0-1-59.node.dc1.consul. 0 IN      A       10.0.1.59
ip-10-0-1-185.node.dc1.consul. 0 IN     A       10.0.1.185
ip-10-0-1-16.node.dc1.consul. 0 IN      A       10.0.1.16
ip-10-0-1-16.node.dc1.consul. 0 IN      A       10.0.1.16

;; Query time: 3 msec
;; SERVER: 127.0.0.1#8600(127.0.0.1)
;; WHEN: Sun Apr 10 05:07:44 UTC 2016
;; MSG SIZE  rcvd: 1172
```

### PR DESCRIPTION
**Pull request by slackpad**
Fixes #1931.

This is more complicated than I would have liked, but because the ADDITIONAL section isn't 1:1 with the ANSWER section, I had to re-derive the ADDITIONAL contents after a trim. I kept an index around to make this as cheap as possible.

### CODE REVIEW

File: command/agent/dns_test.go [hunk 1]
```diff
@@ -1986,8 +1987,8 @@ func TestDNS_ServiceLookup_LargeResponses(t *testing.T) {
 
 	longServiceName := "this-is-a-very-very-very-very-very-long-name-for-a-service"
 
-	// Register 3 nodes
-	for i := 0; i < 3; i++ {
+	// Register a lot of nodes.
+	for i := 0; i < 4; i++ {
```
Comment (sean-) [inline command/agent/dns_test.go]: Why 4 nodes (up from 3)?  To overflow the 512B limit?

Comment (slackpad) [inline command/agent/dns_test.go]: I wanted this to hit the initial trim based on the number of entries, and then hit the size-based trim as well. Just to make it go through more of the code.

File: command/agent/dns.go [hunk 2]
```diff
@@ -494,16 +494,79 @@ func (d *DNSServer) formatNodeRecord(node *structs.Node, addr, qName string, qTy
...
+// records will be trimmed along with answers.
+func trimUDPResponse(config *DNSConfig, resp *dns.Msg) (trimmed bool) {
 	numAnswers := len(resp.Answer)
+	hasExtra := len(resp.Extra) > 0
+
+	// We avoid some function calls and allocations by only handling the
+	// extra data when necessary.
+	var index map[string]dns.RR
+	if hasExtra {
+		index = indexRRs(resp.Extra)
```
Comment (sean-) [inline command/agent/dns.go]: Can you pass `index` in as an argument and modify the value on `trimUDPResponse()`'s stack?

```
index := make(map[string]dns.RR, numAnswers)
...
indexRRs(resp.Extra, &index)
```

This is in the DNS callpath so reducing the number of heap allocations seems important.  I think Go will do if it goes through a function call boundary.  We can measure this during testing and do later, but something I wanted to point out.

Comment (slackpad) [inline command/agent/dns.go]: I went ahead and allocated above and passed by reference.

**Pull request merged**